examples/common.cpp

```diff
@@ -516,14 +514,218 @@ void win32_console_init(bool enable_color) {
+        out.push_back(static_cast<unsigned char>(0x80 | ((ch >> 6) & 0x3F)));
+        out.push_back(static_cast<unsigned char>(0x80 | (ch & 0x3F)));
+    } else if (ch <= 0x10FFFF) {
+        out.push_back(static_cast<unsigned char>(0xF0 | ((ch >> 18) & 0x07)));
+        out.push_back(static_cast<unsigned char>(0x80 | ((ch >> 12) & 0x3F)));
+        out.push_back(static_cast<unsigned char>(0x80 | ((ch >> 6) & 0x3F)));
+        out.push_back(static_cast<unsigned char>(0x80 | (ch & 0x3F)));
+    } else {
+        // Invalid Unicode code point
+    }
+}
+
+// Helper function to remove the last UTF-8 character from a string
+void pop_back_utf8_char(std::string & line) {
+    if (line.empty()) {
+        return;
+    }
+
+    size_t pos = line.length() - 1;
+
+    // Find the start of the last UTF-8 character (checking up to 4 bytes back)
+    for (size_t i = 0; i < 3 && pos > 0; ++i, --pos) {
+        if ((line[pos] & 0xC0) != 0x80) break; // Found the start of the character
+    }
+    line.erase(pos);
+}
+
+bool console_readline(console_state & con_st, std::string & line) {
+    console_set_color(con_st, CONSOLE_COLOR_USER_INPUT);
+
+    line.clear();
+    std::vector<int> widths;
+    bool is_special_char = false;
+    bool end_of_stream = false;
+
+    char32_t input_char;
+    while (true) {
+        fflush(stdout); // Ensure all output is displayed before waiting for input
+        input_char = getchar32();
+
+        if (input_char == '\r' || input_char == '\n') {
+            break;
+        }
+
+        if (input_char == WEOF || input_char == 0x04 /* Ctrl+D*/) {
+            end_of_stream = true;
+            break;
+        }
+
+        if (is_special_char) {
+            console_set_color(con_st, CONSOLE_COLOR_USER_INPUT);
+            putchar('\b');
+            putchar(line.back());
+            is_special_char = false;
+        }
+
+        if (input_char == '\033') { // Escape sequence
+            char32_t code = getchar32();
+            if (code == '[' || code == 0x1B) {
+                // Discard the rest of the escape sequence
+                while ((code = getchar32()) != WEOF) {
+                    if ((code >= 'A' && code <= 'Z') || (code >= 'a' && code <= 'z') || code == '~') {
+                        break;
+                    }
+                }
+            }
+        } else if (input_char == 0x08 || input_char == 0x7F) { // Backspace
+            if (!widths.empty()) {
+                int count;
+                do {
+                    count = widths.back();
+                    widths.pop_back();
+                    // Move cursor back, print spaces, and move cursor back again
+                    for (int i = 0; i < count; i++) {
+                        fputs("\b \b", stdout);
+                    }
+                    pop_back_utf8_char(line);
+                } while (count == 0 && !widths.empty());
+            }
+        } else if (input_char < 32) {
+            // Ignore control characters
```

Comment:
That's what I'm working on right now actually. The tab character doesn't work with wcwidth (returns -1) so I'm trying to find the best solution. I either need to track my position by adding up all the widths (I should be able to read the tab stops using ioctl but it's not a given) or set it to a fixed size. In earlier testing, sending the position reading control sequence to the terminal could break things like combining characters although it may be safe if I'm smart about code points or if I write to the tty directly.

Now that I'm writing this, the tty thing might just be a better solution. It's a bit of added complexity and I'm trying to balance code simplicity with features, but it I guess I'll try it out.

Reply:
I also didn't mention it, but if we wanted to track the cursor position by adding up with widths, it needs more than just the input character widths, we'd need the widths of anything printed.
```C++
void console_printf(console_state & con_st, const char *format, ...) {
    char buf[1024]; // TODO: used vsnprintf to get the required size
    va_list args;

    va_start(args, format);
    vsnprintf(buf, sizeof(buf), format, args);
    va_end(args);

    for (int i = 0; buf[i] != '\0'; i++) {
        if (buf[i] == '\t') {
            int tab_width = 8; // Adjust this value based on your terminal's tab settings
            con_st->cursor_position += tab_width - (con_st->cursor_position % tab_width);
        } else if (buf[i] == '\n') {
            con_st->cursor_position = 0;
        } else {
            con_st->cursor_position += 1; //wcwidth(of code point converted from utf-8);
        }
    }

    printf("%s", buf);
}
```
So it sounds simplest, but is actually a bit more involved.

Reply:
Reading the position from tty isn't as clean as `wcwidth`, but it just works with every single character and combining character.